tsl/src/fdw/estimate.c

```diff
@@ -121,14 +121,26 @@ get_upper_rel_estimate(PlannerInfo *root, RelOptInfo *rel, CostEstimate *ce)
 		get_agg_clause_costs_compat(root, (Node *) fpinfo->grouped_tlist, aggsplit, &aggcosts);
 	}
 
-	/* Get number of grouping columns and possible number of groups */
+	/*
+	 * Get number of grouping columns and possible number of groups. We don't
+	 * have per-column ndistinct statistics on access node for distributed
+	 * tables, so in this case hardcode it as 1/10 of all rows to incentivize
+	 * grouping push down.
+	 */
```

Comment:
The previous estimate was just totally incorrect because of the missing statistics. This improves over it, because GROUP BY should generally produce less rows than the ungrouped relation.

Reply:
We have n_distinct stats for chunks (the root table holds no data so no stats). 

After I run ANALYZE on a distributed hypertable I see this:

```
postgres=# select tablename, attname, n_distinct from pg_stats where tablename LIKE '_dist_hyper_1_16_chunk';
       tablename        | attname | n_distinct 
------------------------+---------+------------
 _dist_hyper_1_16_chunk | time    |         -1
 _dist_hyper_1_16_chunk | device  |          3
 _dist_hyper_1_16_chunk | temp    |         -1
(3 rows)

```
While it makes sense to estimate the n_distinct stats for hypertables when no stats exist, I think we should try to do better with the stats we have.
 
For row count stats we use the sum of all chunks in a data node scan to get the stats for that DN. For n_distinct we could do something similar, but for n_distinct we probably need to use something like the average of all chunks' n_distinct.

If we are calculating the n_distinct for the whole hypertable, we need to look at two cases: (1) the column is the space partitioning column (2) it is not the space partition column.

For case (1) (column is space partition), then we can just sum the n_distinct of all data nodes.

For case (2), we probably need to take the average of all chunks.

Reply:
Sounds reasonable. We would use `get_relation_stats_hook` to return the averaged statistics for a given data node scan into `examine_simple_variable`, and then the Postgres cost estimation functions would make use of it. This looks like a relatively big change though, do you think we can do it later? This PR is already oversized somewhat, I had to make many changes to get good plans. It's better than the status quo, plus it brings us closer to using the parameterized data node scans, which will up the baseline performance for many kinds of multinode joins.

Reply:
This particular place might be simpler, I was thinking about join selectivity estimate in `add_data_node_scan_paths`. I would prefer to implement your suggested approach in both of these places and wherever else it is needed, so that we have a more consistent cost model. I'm going to correct the comments and mention this possibility.